strax/storage/mongo.py

```diff
@@ -11,100 +11,159 @@
+        """See strax.Backend"""
+        query = backend_key_to_query(backend_key)
+        chunk_name = f'chunk_{chunk_info["chunk_i"]}'
+        # Query for the chunk and project the chunk info
+        doc = self.db[self.col_name].find_one(
+            {**query, chunk_name: {"$exists": True}},
+            {f"{chunk_name}.$": 1})
+        doc = doc.get(chunk_name, None)
 
-        doc = self.db[backend_key].find_one(
-            {'chunk': chunk_info['chunk_i']})
         if doc is None:
-            return []
+            return np.array([], dtype=dtype)
 
-        result = np.zeros(len(doc['data']), dtype=dtype)
-        for i, d in enumerate(doc['data']):
+        # Convert JSON to numpy
+        result = np.zeros(len(doc), dtype=dtype)
+        for i, d in enumerate(doc):
             for key in np.dtype(dtype).names:
-                result[i][key] = d[key]
+                result[i][key] = d['data'][key]
         return result
 
     def _saver(self, key, metadata):
-        return MongoSaver(key, metadata, self.db[str(key)])
+        """See strax.Backend"""
+        # Use the key to make a collection otherwise, use the backend-key
+        col = self.db[self.col_name if self.col_name is not None else str(key)]
+        return MongoSaver(key, metadata, col)
 
     def get_metadata(self, key):
-        return self.db[key].find_one({'metadata': True})
+        """See strax.Backend"""
+        query = backend_key_to_query(key)
+        doc = self.db[self.col_name].find_one(query)
+        if doc and 'metadata' in doc:
+            return doc['metadata']
+        raise strax.DataNotAvailable
 
 
 @export
 class MongoFrontend(StorageFrontend):
     """MongoDB storage frontend"""
 
-    def __init__(self, uri, database, *args, **kwargs):
+    def __init__(self, uri, database, col_n=None, *args, **kwargs):
+        """
+        MongoFrontend for reading/writing data from Mongo
+        :param uri: Mongo url (with pw and username)
+        :param database: name of database (str)
+        :param col_n: collection name (str) to look for data
+        :param args: init for StorageFrontend
+        :param kwargs: init for StorageFrontend
+        """
+
         super().__init__(*args, **kwargs)
         if MongoClient is None:
             raise ImportError("Pymongo did not import: "
                               "cannot use MongoDBStore")
         self.client = MongoClient(uri)
         self.db = self.client[database]
-        self.backends = [MongoBackend(uri, database)]
+        self.backends = [MongoBackend(uri, database, col_n=col_n)]
+        self.col_name = col_n
 
     def _find(self, key, write, allow_incomplete, fuzzy_for,
               fuzzy_for_options):
-
+        """See strax.Frontend"""
         if write:
             return self.backends[0].__class__.__name__, str(key)
-        if self.db[str(key)].count():
+        query = backend_key_to_query(str(key))
+        if self.db[self.col_name].find_one(query):
```

Comment:
`self.db[self.col_name].count_documents(query)` might be the more efficient query here.

Reply:
I'm not sure why? Looking at https://docs.mongodb.com/manual/reference/method/db.collection.countDocuments/ it just aggregates the query. I don't mind changing it though.

Reply:
The main reason is that `find_one` returns an entire document, while `count_documents` returns an integer. If the document in question happens to be "large", then the API downloads and returns the entire thing just to see if it exists.

Reply:
Thanks Darryl!

Reply:
I think I've made some occasional similar comments in bootstrax PRs when it calls `find_one_and_update` rather than `update_one` (the difference being whether or not a document is returned), but at least in those cases it's usually much clearer exactly what size the document is that it's downloading and discarding.

Reply:
Thanks, will remember it!